discovery-provider/src/challenges/challenge_event_bus.py

```diff
@@ -0,0 +1,71 @@
+import json
+import enum
+import logging
+
+from collections import defaultdict
+
+logger = logging.getLogger(__name__)
+REDIS_QUEUE_PREFIX = 'challenges-event-queue'
+
+class ChallengeEvent(str, enum.Enum):
+    profile_update = 'profile_update'
+    repost = 'repost'
+    follow = 'follow'
+    favorite = 'favorite'
+
+class ChallengeEventBus:
+    """`ChallengeEventBus` supports:
+        - dispatching challenge events to a Redis queue
+        - registering challenge managers to listen to the events.
+        - consuming items from the Redis queue
+    """
+    def __init__(self, redis):
+        self._listeners = defaultdict(lambda: [])
+        self._redis = redis
+
+    def register_listener(self, event, listener):
+        """Registers a listener (`ChallengeEventManager`) to listen for a particular event type."""
+        self._listeners[event].append(listener)
+
+    def dispatch(self, session, event, block_number, user_id):
+        """Dispatches an event + block_number + user_id to Redis queue"""
+        event_json = self._event_to_json(event, block_number, user_id)
+        logger.debug(f"ChallengeEventBus: dispatch {event_json}")
+        self._redis.rpush(REDIS_QUEUE_PREFIX, event_json)
+
+    def process_events(self, session, max_events=1000):
```

Comment:
slightly concerned about the initial set up cost of doing 1000 at a time w/ a ton of events that we'll have to go through to catch up retroactive users, but i imagine we can flesh that out in subsequent PRs

Reply:
yeah was thinking about this - not sure whether I can just do this all in a migration in SQL, but that's what I'm going to try first. if we have to do redis, there are some other things we can do to make sure we're still processing regular events while we go through the backlog (also, if we could crunch through 1000 events every 5 seconds for example, this would only take 15 min to get through all users)